### PR TITLE
Add cloud functions

### PR DESCRIPTION
**Pull request by gbaranski**

### CODE REVIEW

**Comment by gbaranski** [comment]
Screen Shot 2020-08-08 at 04 19 18
Everything is working, but I've got question @stasbar, how to calculate this things with 'Calculation required', is this a bug?

**Comment by stasbar** [comment]
calculateProperties is for calculating them

File: src/utils/firebase.ts [hunk 1]
```diff
@@ -101,3 +104,22 @@ export function getCurrentUser(): Promise<FirebaseUser | undefined> {
...
+    | 'calculateResults'
+    | 'calculateForMix',
+  data:
+    | { mixable1: Mixable; mixable2: Mixable; preferences?: any }
+    | { coil: Coil; baseVoltage?: number }
+    | { coil: Coil; power: number; heatFlux: number }
+    | { liquid: Liquid },
+): Promise<T> {
+  const res = await functions.httpsCallable(name)({ ...data });
+  console.log({ fbres: res.data });
```
Comment (stasbar) [inline src/utils/firebase.ts:123]: remove console.log

File: src/utils/firebase.ts [hunk 2]
```diff
@@ -101,3 +104,22 @@ export function getCurrentUser(): Promise<FirebaseUser | undefined> {
...
+    | 'calculateForWrapsBasedOnPower'
+    | 'calculateResults'
+    | 'calculateForMix',
+  data:
+    | { mixable1: Mixable; mixable2: Mixable; preferences?: any }
+    | { coil: Coil; baseVoltage?: number }
+    | { coil: Coil; power: number; heatFlux: number }
+    | { liquid: Liquid },
+): Promise<T> {
+  const res = await functions.httpsCallable(name)({ ...data });
```
Comment (stasbar) [inline src/utils/firebase.ts:122]: don't spread single object i.e. `{ ...data } == data`

File: src/services/mixer.ts [hunk 3]
```diff
@@ -1,18 +1,12 @@
...
     if (!auth.currentUser) {
       throw Error('You are not logged in');
     }
-    const idToken = await auth.currentUser!.getIdToken(false);
-    console.log({ data: JSON.stringify({ mixable1, mixable2, idToken }) });
-    return request(`/api/calculator/mixer/mix`, {
-      method: 'POST',
-      data: { mixable1, mixable2, idToken },
-    });
+    return await callFirebaseFunction('calculateForMix', { mixable1, mixable2 });
```
Comment (stasbar) [inline src/services/mixer.ts:9]: don't `return await`, just `return`. Its promise anyway

File: src/services/liquid.ts [hunk 4]
```diff
@@ -1,11 +1,15 @@
...
-export async function calculateResults(liquid: Liquid): Promise<any> {
+export async function calculateResults(liquid: Liquid): Promise<Result[]> {
   if (!auth.currentUser) {
     throw Error('You are not logged in');
   }
-  const idToken = await auth.currentUser!.getIdToken(false);
-  return request('/api/calculator/liquid/results', { method: 'POST', data: { liquid, idToken } });
+  try {
+    const res = await callFirebaseFunction<Result[]>('calculateResults', { liquid });
+    return res;
```
Comment (stasbar) [inline src/services/liquid.ts:10]: inline without await

File: src/services/mixer.ts (hunk 3 above)
Comment (gbaranski) [reply src/services/mixer.ts:9]: @stasbar yes, but try/catch won't catch the error if no await

Comment (stasbar) [reply src/services/mixer.ts:9]: You are right. Didn't notice trycatch

**Pull request merged**